pkg/operator/ceph/cluster/osd/health.go

```diff
@@ -146,3 +151,49 @@ func (m *Monitor) handleOSDMarkedOut(outOSDid int) error {
 	}
 	return nil
 }
+
+// restartOSDIfStuck will check if a portable OSD is on a node that is not ready.
+// If the pod is stuck in terminating state, go ahead and force delete the pod so K8s
+// will free up the volume and allow the OSD to be restarted on another node.
+func (m *Monitor) restartOSDIfStuck(osdID int) error {
+	labels := fmt.Sprintf("ceph-osd-id=%d,portable=true", osdID)
+	pods, err := m.context.Clientset.CoreV1().Pods(m.namespace).List(metav1.ListOptions{LabelSelector: labels})
+	if err != nil {
+		return errors.Wrapf(err, "failed to get OSD with ID %d", osdID)
+	}
+	return m.restartOSDPodsIfStuck(osdID, pods)
+}
+
+func (m *Monitor) restartOSDPodsIfStuck(osdID int, pods *v1.PodList) error {
+	for _, pod := range pods.Items {
+		logger.Debugf("checking if osd %d pod is stuck and should be force deleted", osdID)
+		if pod.DeletionTimestamp == nil {
+			logger.Debugf("skipping restart of OSD %d since the pod is not deleted", osdID)
+			continue
+		}
+		if k8sutil.PodIsReady(pod) {
```

Comment:
If the pod has a deletion timestamp does it matter if it is ready? Could it in a running state on a not-ready node?

Reply:
You're right, this check is really redundant. Checking for the deletion timestamp and then the node status is already covering it.